src/Haskell/Ide/Engine/Scheduler.hs

```diff
@@ -0,0 +1,197 @@
+   -- at a time, this cannot be changed a runtime.
+
+ , requestsToCancel :: STM.TVar (Set.Set J.LspId)
+   -- ^ The request IDs that were canceled by the client. This causes requests to
+   -- not be dispatched or aborted if they are already in progress.
+
+ , requestsInProgress :: STM.TVar (Set.Set J.LspId)
+   -- ^ Requests IDs that have already been dispatched. Currently this is only used to keep
+   -- @requestsToCancel@ bounded. We only insert IDs into the cancel list if the same LspId is
+   -- also present in this variable.
+
+ , documentVersions :: STM.TVar (Map.Map Uri Int)
+   -- ^ A Map containing document file paths with their respective current version. This is used
+   -- to prevent certain requests from being processed if the current version is more recent than
+   -- the version the request is for.
+
+ , ideChan :: (Channel.InChan (IdeRequest m), Channel.OutChan (IdeRequest m))
+   -- ^ Holds the reading and writing ends of the channel used to dispatch Ide requests
+
+ , ghcChan :: (Channel.InChan (GhcRequest m), Channel.OutChan (GhcRequest m))
+   -- ^ Holds the reading and writing ends of the channel used to dispatch Ghc requests
+ }
+
+-- ^ A pair representing the document file path and a new version to store for it.
+type DocUpdate = (Uri, Int)
+
+
+-- | Create a new scheduler parameterized with the monad of your choosing.
+-- This is the monad where the handler for requests and handler for errors will run.
+--
+-- Once created, the scheduler needs to be run using 'runScheduler'
+newScheduler
+  :: IdePlugins
+     -- ^ The list of plugins that will be used for responding to requests
+  -> GM.Options
+   -- ^ Options for the ghc-mod session. Since we only keep a single ghc-mod session
+  -> IO (Scheduler m)
+newScheduler plugins ghcModOptions = do
+  cancelTVar  <- STM.atomically $ STM.newTVar Set.empty
+  wipTVar     <- STM.atomically $ STM.newTVar Set.empty
+  versionTVar <- STM.atomically $ STM.newTVar Map.empty
+  ideChan     <- Channel.newChan
+  ghcChan     <- Channel.newChan
+  return $ Scheduler
+    { plugins            = plugins
+    , ghcModOptions      = ghcModOptions
+    , requestsToCancel   = cancelTVar
+    , requestsInProgress = wipTVar
+    , documentVersions   = versionTVar
+    , ideChan            = ideChan
+    , ghcChan            = ghcChan
+    }
+
+
+-- | Runs the given scheduler. This is meant to run in a separate thread and
+-- the thread should be kept alive as long as you need requests to be dispatched.
+runScheduler
+  :: forall m
+   . Scheduler m
+     -- ^ The scheduler to run.
+  -> ErrorHandler
+     -- ^ A handler for any errors that the dispatcher may encounter.
+  -> CallbackHandler m
+     -- ^ A handler to run the requests' callback in your monad of choosing.
+  -> C.ClientCapabilities
+     -- ^ List of features the IDE client supports or has enabled.
+  -> IO ()
+runScheduler Scheduler {..} errorHandler callbackHandler caps = do
+  let dEnv = DispatcherEnv
```

Comment:
It's great that `cancelTVar`/`wipTVar` are no longer exposed! Could we go one step further and maybe merge Dispatcher/Scheduler so that `DispatcherEnv` is also just an internal structure? It would be really nice if it wasn't exported at all

Reply:
Does any other module other than `Scheduler` import `Dispatcher`?

Reply:
I don't think anything else is importing the dispatcher, I could move the full contents of it into the scheduler now. Do you think I should do that here?

Reply:
That sounds good, we'll probably need to do a little bikeshedding about the naming at some point

Reply:
@bubba I moved all the dispatcher code into the Scheduler, now the DispatcherEnv record is hidden